### PR TITLE
Simple static location sharing

### PR DESCRIPTION
**Pull request by ara4n**
A quick and dirty implementation of plain old `m.location` msgtype location sharing via maplibre-gl, mixing in https://github.com/matrix-org/matrix-doc/pull/3488 for forwards compatibility.

Requires https://github.com/vector-im/element-web/pull/19754

Stuff remaining:
 * [x] fix lint
 * [x] fix types
 * [x] render description
 * [x] stop the UI prompting for streaming until [MSC3489](https://github.com/matrix-org/matrix-doc/pull/3489) streaming is hooked up.
 * [x] fix tests

Stuff punted for future PRs:
 * add thumbnail fallbacks
 * let the user drop a pin in a specific location
 * render uncertainty circle

---
Here's what your changelog entry will look like:

## ✨ Features
 * Simple static location sharing ([\#7135](https://github.com/matrix-org/matrix-react-sdk/pull/7135)).

Preview: https://61acb1cab46d1e0ed42b507a--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

**Comment by ara4n** [comment]
Other than needing to figure out how to fix the tests by mocking maplibre correctly (maybe using https://github.com/mapbox/mapbox-gl-js-mock?) i think this is good to go.

File: src/components/views/location/LocationShareType.tsx [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    ONE_OFF = 0,
+    ONE_MIN = 60,
+    FIVE_MINS = 5 * 60,
+    THIRTY_MINS = 30 * 60,
+    ONE_HOUR = 60 * 60,
+    THREE_HOURS = 3 * 60 * 60,
+    SIX_HOURS = 6 * 60 * 60,
+    ONE_DAY = 24 * 60 * 60,
+    FOREVER = Number.MAX_SAFE_INTEGER,
+}
```
Comment (t3chguy) [inline src/components/views/location/LocationShareType.tsx:28]: We tend to use PascalCasing for enum members

File: res/css/views/location/_LocationPicker.scss [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    height: 500px;
+
+    border-radius: 4px;
+
+    display: flex;
+    flex-direction: column;
+}
+
+#mx_LocationPicker_map {
+    height: 324px;
```
Comment (t3chguy) [inline res/css/views/location/_LocationPicker.scss:28]: this seems very magical, whats the justification?

File: res/css/views/location/_LocationPicker.scss [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    border-radius: 8px 8px 0px 0px;
+}
+
+.mx_LocationPicker_footer {
+    margin: 24px;
+    // border-top: 1px solid $message-action-bar-border-color;
+    // min-height: 72px;
+
+    // display: flex;
+    // align-items: center;
```
Comment (t3chguy) [inline res/css/views/location/_LocationPicker.scss:38]: what's with these?

File: src/components/views/location/LocationPicker.tsx [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+interface IProps {
+    onChoose(uri: string, ts: number, type: LocationShareType, description: string): boolean;
+    onFinished();
+}
+
+interface IState {
+    description: string;
+    type: LocationShareType;
+    position: GeolocationPosition;
```
Comment (t3chguy) [inline src/components/views/location/LocationPicker.tsx:75]:
```suggestion
    position?: GeolocationPosition;
```

This gets set to undefined in the c'tor

File: src/components/views/messages/MLocationBody.tsx [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+import maplibregl from 'maplibre-gl';
+import SdkConfig from '../../../SdkConfig';
+import { replaceableComponent } from "../../../utils/replaceableComponent";
+import { IBodyProps } from "./IBodyProps";
+
+interface IState {
+}
+
+@replaceableComponent("views.messages.MLocationBody")
+export default class MLocationBody extends React.Component<IBodyProps, IState> {
```
Comment (t3chguy) [inline src/components/views/messages/MLocationBody.tsx:27]:
```suggestion
@replaceableComponent("views.messages.MLocationBody")
export default class MLocationBody extends React.Component<IBodyProps> {
```

Stateless components shouldn't pass any IState

File: src/components/views/messages/MLocationBody.tsx [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+            content['geo_uri'];
+
+        this.coords = this.parseGeoUri(uri);
+
+        this.description =
+            (content['org.matrix.msc3488.location'] &&
+                content['org.matrix.msc3488.location'].description !== undefined) ?
+                content['org.matrix.msc3488.location'].description :
+                content['body'];
+    }
```
Comment (t3chguy) [inline src/components/views/messages/MLocationBody.tsx:51]:
```suggestion
        this.description =
            content['org.matrix.msc3488.location']?.description ?? content['body'];
    }
```

File: src/components/views/rooms/MessageComposer.tsx [hunk 7]
```diff
@@ -126,6 +130,46 @@ const EmojiButton: React.FC<IEmojiButtonProps> = ({ addEmoji, menuPosition, narr
     </React.Fragment>;
 };
 
+interface ILocationButtonProps {
+    room: Room;
+    shareLocation: (uri: string, ts: number, type: LocationShareType, description: string) => boolean;
+    menuPosition: any; // TODO: Types
```
Comment (t3chguy) [inline src/components/views/rooms/MessageComposer.tsx:136]: `AboveLeftOf` is the type for this, from ContextMenu

**Review by t3chguy — APPROVED** [review]
Looks sane otherwise assuming tests are fixed

**Comment by turt2live** [comment]
This is blocked on internal discussion. Please ping me for context.

**Review by turt2live — DISMISSED** [review]
I suppose I should use this feature to block merging as well...

File: res/css/views/location/_LocationPicker.scss (hunk 2 above)
Comment (ara4n) [reply res/css/views/location/_LocationPicker.scss:28]: "it looks pretty". will be replaced once we have an actual design.

File: res/css/views/location/_LocationPicker.scss (hunk 3 above)
Comment (ara4n) [reply res/css/views/location/_LocationPicker.scss:38]: oops, cargoculted and forgot to remove them.

**Comment by SimonBrandner** [comment]
Screenshot_20211123_074726

Is it expected for me to see an empty map?

**Review by andybalaam — APPROVED** [review]
Looks good to me.

**Comment by t3chguy** [comment]
@SimonBrandner I'm guessing its due to missing API in config.json

**Review by t3chguy — CHANGES_REQUESTED** [review]
image

- [ ] Should not be enabled/error when the API is missing in config.json
- [ ] Should handle API errors better than just showing no map.

**Comment by andybalaam** [comment]
Confirming this worked for me on a local install with a config file containing a maptiler API key.

File: src/components/views/messages/MLocationBody.tsx [hunk 8]
```diff
@@ -84,8 +97,14 @@ export default class MLocationBody extends React.Component<IBodyProps> {
     };
 
     render() {
+        const error = this.state.fail ?
+            <div className="mx_EventTile_tileError mx_EventTile_body">
+                Failed to load map - check config.json has a valid API key
```
Comment (t3chguy) [inline src/components/views/messages/MLocationBody.tsx:102]: needs i18n

Comment (t3chguy) [reply src/components/views/messages/MLocationBody.tsx:102]: also probably not a great user-facing error given majority of users won't be in control of the config.json, this should probably go into the console with a glossier user-facing error

**Comment by ara4n** [comment]
This now checks and warns for misconfigured tileserver nicely. PTAL; should be good to merge now.

File: res/css/views/location/_LocationPicker.scss [hunk 9]
```diff
@@ -37,3 +37,8 @@ limitations under the License.
         max-height: 140px;
     }
 }
+
+.mx_LocationPicker_error {
+    color: red;
```
Comment (t3chguy) [inline res/css/views/location/_LocationPicker.scss:42]: this should be a themed colour

**Review by t3chguy — APPROVED** [review]
:shipit:

File: res/css/views/location/_LocationPicker.scss (hunk 9 above)
Comment (ara4n) [reply res/css/views/location/_LocationPicker.scss:42]: all the other red errors are hardcoded currently, and meanwhile there are a tonne of hardcoded colours flying around:

```
matthew@shadowfax css % egrep -r 'color: (#|[^\$])' | wc -l
     126
```

...so given it's an edge error case, i'm inclined to leave it be and have de-hardcoding colours as a separate PR for someone in future.

**Comment by poiNt3D** [comment]
Can't find the button.

**Comment by turt2live** [comment]
Please use the community support rooms, like [#element-web:matrix.org](https://matrix.to/#/#element-web:matrix.org) for support.

**Pull request merged**